### PR TITLE
Find unused TCP port - HttpInvoker Part 4

### PR DESCRIPTION
**Pull request by pragnagopa**
- Find unused TCP port to host http invoker service
- Pass command line `--httpInvokerPort portNumber` to invoker service

Related issue #https://github.com/Azure/azure-functions-host/issues/4930

Part 4 for PR #https://github.com/Azure/azure-functions-host/pulls/pragnagopa

### CODE REVIEW

**Comment by pragnagopa** [comment]
@fabiocav / @mhoeger /@yojagad - please review. This is part 4 of PR#Azure#4920

**Comment by pragnagopa** [comment]
Sample HttpInvokerService in go: https://github.com/pragnagopa/functions-http-worker/blob/master/go/src/httpServer/helloworldServer.go

File: src/WebJobs.Script/OutOfProc/Http/Configuration/HttpInvokerOptionsSetup.cs [hunk 1]
```diff
@@ -42,8 +44,19 @@ public void Configure(HttpInvokerOptions options)
...
+
+        internal static int GetUnusedTcpPort()
+        {
+            // TODO verify in Azure
+            TcpListener tcpListener = new TcpListener(IPAddress.Loopback, 0);
+            tcpListener.Start();
+            int port = ((IPEndPoint)tcpListener.LocalEndpoint).Port;
+            tcpListener.Stop();
+            return port;
+        }
```
Comment (yojagad) [inline src/WebJobs.Script/OutOfProc/Http/Configuration/HttpInvokerOptionsSetup.cs:60]: If a port is found here, and then the listener is not bound to it anymore, there is a chance (miniscule?) that the port number found here could be bound to something else before it goes to the _processFactory?

File: src/WebJobs.Script/OutOfProc/Rpc/RpcInitializationService.cs [hunk 2]
```diff
@@ -164,6 +175,14 @@ internal bool ShouldStartInPlaceholderMode()
             return false;
         }
 
+        internal bool ShouldStartAsPlaceholderPool()
+        {
+            // We are in placeholder mode but a worker runtime IS set
+            return _environment.IsPlaceholderModeEnabled()
+                && !string.IsNullOrEmpty(_workerRuntime)
+                && _placeholderPoolWhitelistedRuntimes.Contains(_workerRuntime);
```
Comment (yojagad) [inline src/WebJobs.Script/OutOfProc/Rpc/RpcInitializationService.cs]: >_placeholderPoolWhitelistedRuntimes.Contains(_workerRuntime); [](start = 19, length = 61)

Do we need to worry about casing here?

File: src/WebJobs.Script/OutOfProc/Rpc/WebHostLanguageWorkerChannelManager.cs [hunk 3]
```diff
@@ -102,16 +102,42 @@ public async Task SpecializeAsync()
...
         }
 
+        private bool UsePlaceholderChannel(string workerRuntime)
+        {
+            if (!string.IsNullOrEmpty(workerRuntime))
+            {
+                // Special case: node apps must be read-only to use the placeholder mode channel
+                // TODO: Remove special casing when resolving https://github.com/Azure/azure-functions-host/issues/4534
+                if (string.Equals(workerRuntime, LanguageWorkerConstants.NodeLanguageWorkerName))
+                {
```
Comment (yojagad) [inline src/WebJobs.Script/OutOfProc/Rpc/WebHostLanguageWorkerChannelManager.cs]: Same comment. Do we need to worry about casing?

File: src/WebJobs.Script/OutOfProc/Http/Configuration/HttpInvokerOptionsSetup.cs (hunk 1 above)
Comment (pragnagopa) [reply src/WebJobs.Script/OutOfProc/Http/Configuration/HttpInvokerOptionsSetup.cs:60]: true! but host as of now does not open any other ports so should be ok

File: src/WebJobs.Script/OutOfProc/Rpc/RpcInitializationService.cs (hunk 2 above)
Comment (pragnagopa) [reply src/WebJobs.Script/OutOfProc/Rpc/RpcInitializationService.cs]: apologies. this looks like a merge commit. But good catch we should fix this in dev branch. Will send a PR. FYI @mhoeger

File: src/WebJobs.Script/OutOfProc/Rpc/WebHostLanguageWorkerChannelManager.cs (hunk 3 above)
Comment (pragnagopa) [reply src/WebJobs.Script/OutOfProc/Rpc/WebHostLanguageWorkerChannelManager.cs]: same here. this change is in dev already. we need to fix it

File: src/WebJobs.Script/OutOfProc/Http/Configuration/HttpInvokerOptionsSetup.cs (hunk 1 above)
Comment (yojagad) [reply src/WebJobs.Script/OutOfProc/Http/Configuration/HttpInvokerOptionsSetup.cs:60]: The OS could still assign that port to something else meanwhile, right?

Comment (pragnagopa) [reply src/WebJobs.Script/OutOfProc/Http/Configuration/HttpInvokerOptionsSetup.cs:60]: yes. possible. As of now, I could not find a better way to do this.

**Comment by pragnagopa** [comment]
addressed feedback on this PR. Closing in favor of: #6

**Pull request closed without merging**